storage/session.go

```diff
@@ -0,0 +1,172 @@
+/*
+ * Copyright (C) 2023 Nuts community
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+
+package storage
+
+import (
+	"context"
+	"encoding/json"
+	"github.com/nuts-foundation/nuts-node/storage/log"
+	"strings"
+	"sync"
+	"time"
+)
+
+var _ SessionDatabase = (*inMemorySessionDatabase)(nil)
+var _ SessionStore = (*inMemorySessionStore)(nil)
+
+var sessionStorePruneInterval = 10 * time.Minute
+
+type expiringEntry struct {
+	// Value stores the actual value as JSON
+	Value  string
+	Expiry time.Time
+}
+
+// SessionDatabase is an in memory database that holds session data on a KV basis.
+// Keys could be access tokens, nonce's, authorization codes, etc.
+// All entries are stored with a TTL, so they will be removed automatically.
+type inMemorySessionDatabase struct {
+	cancel   context.CancelFunc
+	ctx      context.Context
+	mux      sync.RWMutex
+	routines sync.WaitGroup
+	entries  map[string]expiringEntry
+}
+
+// NewInMemorySessionDatabase creates a new in memory session database.
+func NewInMemorySessionDatabase() SessionDatabase {
+	result := &inMemorySessionDatabase{
+		entries: map[string]expiringEntry{},
+	}
+	result.ctx, result.cancel = context.WithCancel(context.Background())
```

Comment:
> Don't add a Context member to a struct type; instead add a ctx parameter to each method on that type that needs to pass it along. [(source)](https://github.com/golang/go/wiki/CodeReviewComments#contexts)

If the context is not passed in to the constructor (and not connected to anything), and the context.Err() is not returned from `startPruning`, the same can probably be a achieved with a `done = make(chan struct,0)`?

Reply:
Ideally a context is passed through the Start method of the engine..